### ISSUE TITLE
Test if Requests can request the same URL or URL byte range multiple times

### ISSUE THREAD

**Issue by teor2345** [pastly/simple-bw-scanner#125]
Following on from #123 and #124, we can host a much smaller file on the HTTPS server if:
1. Requests can ask for the same file several times without a significant delay.

It would also be helpful if:
2. Requests can ask for the whole file, and then ask for a byte range in the file.

This would allow us to use the existing Torflow files.

For example:
1G = 16 * 64M
100M = 64M + 64M[0:46M]

We should make sure we disable HTTP compression if we use the same file, otherwise requests like 10000 * 1k might be unusually fast.

**Comment by teor2345** [comment]
If we decide to implement large requests as multiple requests for the same URL, then we should also add an option min_file_size, which is the actual remote minimum file size. Then we can make the max_download_size checks from #124 depend on it instead.

**Comment by pastly** [comment]
I would greatly prefer the simplicity of requiring the user to find/make destinations that have files available that are `max_download_size` or larger. I find it hard to believe 1 GiB, the default `max_download_size`, is too big of a file for bwauths to put on a webserver, but I could just be ignorant. 

To directly address this ticket:

1. Yes HTTP allows you to request something many times before you start reading any response. 
2. Yes HTTP allows you to request something, read it all, request it again on the same connection, read it all, [... repeat until satisfied]. 

But regarding (1): No I don't think Requests makes this easy.

And regarding (2): I do this already,  but **between** measurements. I think the added RTTs **within** a measurement are suboptimal and will hurt result consistency/accuracy.

**Comment by teor2345** [comment]
The Range header coalesces overlapping byte ranges, so we can't request the same bytes twice:
https://tools.ietf.org/html/rfc7233

Regarding (1), the workflow always involves reading the response headers:
http://docs.python-requests.org/en/master/user/advanced/#body-content-workflow

Perhaps asynchronous IO is the way to go?
http://docs.python-requests.org/en/master/user/advanced/#blocking-or-non-blocking

But I think you should just define a minimum file size as part of the protocol.

**Comment by pastly** [comment]
> But I think you should just define a minimum file size as part of the protocol.

Agreed. Well ... it's not part of the **protocol**, necessarily, but there's a lower limit on what the file size can be, **configurable** by the user (defaulting to 1 GiB). The option name is `max_download_size`.

So with that, I'm considering this implemented and ready to close once my http code is merged.
